### ISSUE TITLE
"Block-scoped variable used before its declaration" not detected in some cases

### ISSUE THREAD

**Issue by T18970237136** [Microsoft/TypeScript#7477]
Hi, this is probably only a minor/cosmetic issue, but while trying the code in https://bugzilla.mozilla.org/show_bug.cgi?id=1069480, I found that the Typescript Compiler does not detect that a variable is used before its initialization in a for-of loop with object destructuring using a computed property name.

**TypeScript Version:** 
master (a75a02c)

**Code** ([Playground](http://www.typescriptlang.org/Playground#src=%22use%20strict%22%3B%0A%0A%2F%2F%20https%3A%2F%2Fbugzilla.mozilla.org%2Fshow_bug.cgi%3Fid%3D1069480%0A%0A%2F%2F%201%3A%0Afor%20%28let%20{[a]%3A%20a}%20of%20[{%20}]%29%20continue%3B%0A%0A%2F%2F%202%3A%0Afor%20%28let%20{[a]%3A%20a}%20%3D%20{%20}%3B%20false%3B%20%29%20continue%3B%0A%0A%2F%2F%203%3A%0Alet%20{[b]%3A%20b}%20%3D%20{%20}%3B))
```ts
"use strict";

// 1:
for (let {[a]: a} of [{ }]) continue;

// 2:
for (let {[a]: a} = { }; false; ) continue;

// 3:
let {[b]: b} = { };
```

**Expected behavior:**
For 1), 2) and 3) the error "Block-scoped variable used before its declaration" should be reported.

**Actual behavior:** 
No error reported for 1), correct errors reported for 2) and 3).

Thanks!
